lib/imgbundler/imgbundler.go

```diff
@@ -6,156 +6,202 @@ import (
-		close(respChan)
-	}()
-
 	for {
 		select {
 		case <-ctx.Done():
-			return nil, fmt.Errorf("failed to wait for imgbundler workers: %w", ctx.Err())
+			return svg, xerrors.Errorf("failed to wait for workers: %w", ctx.Err())
 		case <-time.After(time.Second * 5):
 			ms.Log.Info.Printf("fetching images...")
-		case resp, ok := <-respChan:
+		case repl, ok := <-replc:
 			if !ok {
-				return svg, err
-			}
-			if resp.err != nil {
-				err = multierr.Combine(err, resp.err)
-				continue
+				if len(errhrefs) > 0 {
+					return svg, xerrors.Errorf("%v", errhrefs)
+				}
+				return svg, nil
 			}
-			svg = bytes.Replace(svg, []byte(resp.srctxt), []byte(fmt.Sprintf(`<image href="%s"`, resp.data)), 1)
+			svg = bytes.Replace(svg, repl.from, repl.to, 1)
 		}
 	}
 }
 
-var imgClient = &http.Client{}
+func worker(ctx context.Context, href []byte, isRemote bool) ([]byte, error) {
+	var buf []byte
+	var mimeType string
+	var err error
+	if isRemote {
+		buf, mimeType, err = httpGet(ctx, string(href))
+	} else {
+		buf, err = os.ReadFile(string(href))
+	}
+	if err != nil {
+		return nil, err
+	}
 
-func fetch(ctx context.Context, href string) (string, error) {
+	if mimeType == "" {
+		mimeType = sniffMimeType(href, buf, isRemote)
+	}
+	mimeType = strings.Replace(mimeType, "text/xml", "image/svg+xml", 1)
+	b64 := base64.StdEncoding.EncodeToString(buf)
+	return []byte(fmt.Sprintf(`<image href="data:%s;base64,%s"`, mimeType, b64)), nil
+}
+
+var httpClient = &http.Client{}
+
+func httpGet(ctx context.Context, href string) ([]byte, string, error) {
 	ctx, cancel := context.WithTimeout(ctx, time.Minute)
 	defer cancel()
 
 	req, err := http.NewRequestWithContext(ctx, "GET", href, nil)
 	if err != nil {
-		return "", err
+		return nil, "", err
 	}
 
-	imgResp, err := imgClient.Do(req)
+	resp, err := httpClient.Do(req)
 	if err != nil {
-		return "", err
+		return nil, "", err
 	}
-	defer imgResp.Body.Close()
-	if imgResp.StatusCode != 200 {
-		return "", fmt.Errorf("img %s returned status code %d", href, imgResp.StatusCode)
+	defer resp.Body.Close()
+	if resp.StatusCode != 200 {
+		return nil, "", fmt.Errorf("expected status 200 but got %d %s", resp.StatusCode, resp.Status)
 	}
-	r := http.MaxBytesReader(nil, imgResp.Body, maxImageSize)
-	data, err := ioutil.ReadAll(r)
+	r := http.MaxBytesReader(nil, resp.Body, maxImageSize)
+	buf, err := ioutil.ReadAll(r)
 	if err != nil {
-		return "", err
+		return nil, "", err
 	}
-
-	mimeType := http.DetectContentType(data)
-	mimeType = strings.Replace(mimeType, "text/xml", "image/svg+xml", 1)
-
-	enc := base64.StdEncoding.EncodeToString(data)
-
-	return fmt.Sprintf("data:%s;base64,%s", mimeType, enc), nil
+	return buf, resp.Header.Get("Content-Type"), nil
 }
 
-func read(href string) (string, error) {
-	data, err := os.ReadFile(href)
-	if err != nil {
-		return "", err
+// sniffMimeType sniffs the mime type of href based on its file extension and contents.
+func sniffMimeType(href, buf []byte, isRemote bool) string {
```

Comment:
nice

Reply:
Yea the github image was being detected as text/plain for me lol.